py4DSTEM/process/preprocess/preprocess.py

```diff
@@ -149,5 +150,117 @@ def bin_data_real(datacube, bin_factor):
+        np.roll(diff_mean,( 0, 0),axis=(0,1)).ravel(),
+        np.roll(diff_mean,( 1, 0),axis=(0,1)).ravel(),
+        np.roll(diff_mean,(-1, 1),axis=(0,1)).ravel(),
+        np.roll(diff_mean,( 0, 1),axis=(0,1)).ravel(),
+        np.roll(diff_mean,( 1, 1),axis=(0,1)).ravel(),
+        \
+        np.roll(diff_mean,(-1,-2),axis=(0,1)).ravel(),
+        np.roll(diff_mean,( 0,-2),axis=(0,1)).ravel(),
+        np.roll(diff_mean,( 1,-2),axis=(0,1)).ravel(),
+        np.roll(diff_mean,(-1, 2),axis=(0,1)).ravel(),
+        np.roll(diff_mean,( 0, 2),axis=(0,1)).ravel(),
+        np.roll(diff_mean,( 1, 2),axis=(0,1)).ravel(),
+        \
+        np.roll(diff_mean,(-2,-1),axis=(0,1)).ravel(),
+        np.roll(diff_mean,(-2, 0),axis=(0,1)).ravel(),
+        np.roll(diff_mean,(-2, 1),axis=(0,1)).ravel(),
+        np.roll(diff_mean,( 2,-1),axis=(0,1)).ravel(),
+        np.roll(diff_mean,( 2, 0),axis=(0,1)).ravel(),
+        np.roll(diff_mean,( 2, 1),axis=(0,1)).ravel(),        
+        ]), axis=0)
+    diff_compare = np.reshape(diff_local_med[-ind_compare-1,:], diff_mean.shape)
+
+    # Generate mask
+    mask = diff_mean - diff_compare > thresh
+
+    # apply filtering
+    for ax, ay in tqdmnd(*(datacube.R_Nx,datacube.R_Ny), desc="Cleaning pixels", unit=" images"):
+        # Calculate local 3x3 median images
+        im_med = median_filter(
+            datacube.data[ax,ay,:,:],
+            size=3,
+            mode='nearest')
+        datacube.data[ax,ay,:,:][mask] = im_med[mask]
+
+    if return_mask is True:
+        return datacube, mask
+    else:
+        return datacube
 
+def datacube_diffraction_shift(
+    datacube,
+    xshifts,
+    yshifts,
+    periodic=True,
+    bilinear=False,
+    ):
+    """
+    This function shifts each 2D diffraction image by the values defined by (xshifts,yshifts). 
+    The shift values can be scalars (same shift for all images) or arrays with the same dimensions as
+    the probe positions in datacube.
 
+    Args:
+            datacube (DataCube):   py4DSTEM DataCube
+            xshifts (float):       Array or scalar value for the x dim shifts
+            yshifts (float):       Array or scalar value for the y dim shifts
+            periodic (bool):       Flag for periodic boundary conditions. If set to false, boundaries are assumed to be periodic.
+            bilinear (bool):       Flag for bilinear image shifts. If set to False, Fourier shifting is used.
+
+        Returns:
+            datacube (DataCube):   py4DSTEM DataCube              
+    """
+
+    # if the shift values are constant, expand to arrays
+    xshifts = np.array(xshifts)
+    yshifts = np.array(yshifts)
+    if xshifts.ndim == 0:
+        xshifts = xshifts * np.ones((datacube.R_Nx,datacube.R_Ny))
+    if yshifts.ndim == 0:
+        yshifts = yshifts * np.ones((datacube.R_Nx,datacube.R_Ny))    
+
+    # Loop over all images
+    for ax, ay in tqdmnd(*(datacube.R_Nx,datacube.R_Ny), desc="Shifting images", unit=" images"):
```

Comment:
I'm a little confused why the `*` operator is needed here? maybe we should add a docstring to the tqdmnd function

Reply:
I don't know what the * does, I copied it from some code from @sezelt

Reply:
meh, ok, good enough for now.  the * should unpack the tuple, so that the `tqdmnd` function receives each of those arguments within that tuple individually, i think.  but i don't understand why this is what we're doing, b/c tqdmnd has no documentation, so i'm not really sure what it expect.  so that's mostly what i was complaining about.....not actually related to this PR!